build_tools/cmake/iree_bytecode_module.cmake

```diff
@@ -62,14 +62,29 @@ function(iree_bytecode_module)
     set(_COMPILE_TOOL "iree-compile")
   endif()
 
+  iree_get_executable_path(_COMPILE_TOOL_EXECUTABLE ${_COMPILE_TOOL})
+
+  # Extend _RULE_DEPENDS with the compiler library implementation so changes
+  # to the compiler dynamic library trigger rebuilding bytecode modules.
+  #
+  # There is a dependency chain between:
+  #     code -> static library -> shared library -> tools -> generated files
+  # * On Linux, code changes link the shared library _and_ the tools, which
+  #   triggers rebuilding generated files
+  # * On Windows, code changes link _just_ the shared library, which does _not_
+  #   trigger rebuilding generated files
+  # This makes the dependency explicit, at the potential cost of extra rebuilds.
+  #
+  # Note: Other COMPILE_TOOLs may use different implementations
+  list(APPEND _DEPENDS "${_RULE_DEPENDS}")
+  list(APPEND _DEPENDS "iree::compiler::API2::SharedImpl")
```

Comment:
Hmmm... https://github.com/iree-org/iree/actions/runs/3597840601/jobs/6060089880

```
+ /usr/bin/cmake --build build-riscv-rv32-linux -- -k 0
[0/2] Re-checking globbed directories...
ninja: error: 'runtime/src/iree/hal/drivers/local_sync/cts/iree::compiler::API2::SharedImpl', needed by 'runtime/src/iree/hal/drivers/local_sync/cts/llvm-cpu_command_buffer_dispatch_test.bin', missing and no known rule to make it
```

This isn't going to work for cross compilation, I might need to point at the .dll also, or find a way to re-link (or just touch) the .exe when the .dll changes.

Reply:
I guess I could copy what `iree_get_executable_path` does and depend on `${IREE_HOST_BINARY_ROOT}/bin/IREECompiler.dll`. There's a generator expression I could use: `$<TARGET_RUNTIME_DLLS:iree-compile>`, but that won't work when cross compiling either...

Reply:
Searching for a way to generalize this and avoid hacking something in...

There are also ways to check which DLLs a given executable depends on: https://stackoverflow.com/questions/7378959/how-to-check-for-dll-dependency but I don't think any of those are fully portable to any Windows dev environment... though maybe we could assume a Visual Studio Command Prompt (building with MSVC)?

So we could do something like
* `dumpbin /dependents ${_EXECUTABLE_PATH}`
* regex search output for `IREE*`, find `IREECompiler.dll` (we also depend on `KERNEL32.dll`, `VCRUNTIME140.dll`, etc.)
* add dependency on `${IREE_HOST_BINARY_ROOT}/bin/IREECompiler.dll`

---

The more I think in that direction the more I think it would be easier if `iree-compile.exe` just changed whenever `IREECompiler.dll` changed (either via linking / link flags or some other mechanism)